src/dvc_data/index/save.py

```diff
@@ -21,13 +21,13 @@ def md5(index: "BaseDataIndex", state: Optional["StateBase"] = None) -> None:
     from ..hashfile.hash import fobj_md5
 
     for _, entry in index.iteritems():
-        assert entry.fs
         if entry.meta and entry.meta.isdir:
             continue
 
         if entry.hash_info and entry.hash_info.name == "md5":
             continue
 
+        assert entry.fs
```

Comment:
`entry.fs` will not always be set for intermediate subdir keys, and we don't need to check for a valid `entry.fs` until we actually try to compute md5 on a file

(This assertion is done for mypy/type checking reasons)